course/migrations/0112_add_previous_completion_time_to_flow_session.py

```diff
@@ -0,0 +1,34 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.9 on 2018-01-24 09:04
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+
+
+def set_default_previous_completion_time(apps, schema_editor):
+    FlowSession = apps.get_model("course", "FlowSession")  # noqa
+    target_flow_sessions = FlowSession.objects.filter(
+        in_progress=False
+    )
+    for fs in target_flow_sessions:
+        fs.previous_completion_time = fs.previous_completion_time
+        fs.save()
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('course', '0111_alter_git_source_in_course_to_a_required_field'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='flowsession',
+            name='previous_completion_time',
+            field=models.DateTimeField(blank=True,
+                                       help_text='The last completion time for an in-progress flow session, which had been submitted and then reopened.',
+                                       null=True,
+                                       verbose_name='Previous completion time'),
+        ),
+        migrations.RunPython(set_default_previous_completion_time)
+    ]
```

Comment:
This is a shorthand way for judging whether a session is reopened, and make it easier for sorting GradeChanges with a session.

Reply:
> make it easier for sorting session-related GradeChanges.

To my mind, you should be able to evaluate a grade without having to bring in any associated `FlowSession`s.

Reply:
I think a better path for this would be to have an `effective_time` and a `creation_time` for `GradeChange`s. The `effective_time` would then match the last completion time of an event.

Reply:
As discussed in #430, we need student/grader to know the latest session is reopened, there will be no grade for the opportunity before the reopened session is resubmitted and the resubmitted is graded.

To achieve that, we need to differentiate in-progress sessions between ones which has previously been submitted and then reopened, and ones which hasn't. That can be done by introduction `previous_completion_time` field, because for in-progress sessions which had never been submitted, that field is None. 

This also make it possible to find out which session is the actually the latest session (for grading strategy "use_latest"). Because (IMO and you might also agree), the latest is the session which is submitted latest. Without this, when a latest session is reopened, it is not the latest session any longer. That's a wrong behavior.

Reply:
I used it here https://github.com/inducer/relate/pull/423/files#diff-5a7d8b42274c6175f980693ada70368bR1749

Reply:
What do you think about this? I think this issue is somewhat urgent for me.

Reply:
Can you reply on this discussion? @inducer 
This fix is quite urgent for me. If that bug continued without fixing, some students will definitely complain about their final points. And I won't have time to do more coding.

> I think a better path for this would be to have an effective_time and a creation_time for GradeChanges. The effective_time would then match the last completion time of an event.

Since you seemed to agree on #417 , without adding a `previous_completion_time` field into `Flowsession`, how can we know the session is a reopened one (instead of a just-started on ), and the GradeChange is from a reopened session?

Reply:
I just realized that finishing a session will create a gradechange object, whether autograded or not. I now think your proposal is more reasonable than mine.